include/wlr/types/wlr_pointer_constraints_v1.h

```diff
@@ -0,0 +1,115 @@
+#ifndef WLR_TYPES_WLR_POINTER_CONSTRAINTS_V1_H
+#define WLR_TYPES_WLR_POINTER_CONSTRAINTS_V1_H
+
+#include <stdint.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_box.h>
+#include <wlr/types/wlr_seat.h>
+
+struct wlr_seat;
+
+enum wlr_pointer_constraint_v1_type {
+	WLR_POINTER_CONSTRAINT_V1_LOCKED,
+	WLR_POINTER_CONSTRAINT_V1_CONFINED,
+};
+
+struct wlr_pointer_constraint_v1_state {
+	struct wl_resource *region;
+
+	// only valid for locked_pointer
+	struct {
+		double x, y;
+		bool valid;
+	} cursor_hint;
+};
+
+struct wlr_pointer_constraint_v1 {
+	struct wlr_pointer_constraints_v1 *pointer_constraints;
+
+	struct wl_resource *resource;
+	struct wlr_surface *surface;
+	struct wlr_seat_client *seat_client;
+	uint32_t lifetime;
+
+	bool active;
+	bool accepted;
+
+	struct wlr_pointer_constraint_v1_state current, pending;
+
+	enum wlr_pointer_constraint_v1_type type;
+
+	struct wl_listener surface_commit;
+	struct wl_listener surface_destroy;
+
+	struct wl_list link; // wlr_surface::pointer_constraints
+
+	void *data;
+};
+
+struct wlr_pointer_constraint_v1_activation {
+	struct wlr_pointer_constraint_v1 *pointer_constraint;
+	struct wlr_box *box;
+};
+
+struct wlr_pointer_constraints_v1 {
+	struct wl_global *wl_global;
+
+	struct {
+		/**
+		 * Called when a constraint does not fulfill the requirements
+		 * anymore.
+		 *
+		 * Never called twice in a row.
+		 *
+		 * data: wlr_pointer_constraint_v1*
+		 */
+		struct wl_signal constraint_inactive;
+		/**
+		 * Called when a constraint fulfills all requirements and is
+		 * activated.
+		 *
+		 * Never called twice in a row.
+		 *
+		 * data: wlr_pointer_constraint_v1_activation* (lives on stack)
+		 */
+		struct wl_signal constraint_active;
+		/**
+		 * Called when a constraint is requested by a client.
+		 * The constrain may not fulfill the requirements when
+		 * this is called.
+		 *
+		 * If you wish to accept the request,
+		 * you must call wlr_pointer_constraint_v1_accept on it.
+		 *
+		 * NB: You can call it at any time, so you can e.g.
+		 * first require some specific user input, then call it.
+		 *
+		 * data: wlr_pointer_constraint_v1*
+		 */
+		struct wl_signal request_constraint;
+	} events;
+
+	void* data;
+};
+
+struct wlr_pointer_constraints_v1 *wlr_pointer_constraints_v1_create(
+	struct wl_display *display);
+void wlr_pointer_constraints_v1_destroy(
+	struct wlr_pointer_constraints_v1 *wlr_pointer_constraints_v1);
+
+void wlr_pointer_constraint_v1_accept(
+	struct wlr_pointer_constraint_v1 *constraint);
+void wlr_pointer_constraint_v1_decline(
+	struct wlr_pointer_constraint_v1 *constraint);
+
+// Internal function
+void wlr_pointer_constraints_v1_check_constraint_active(
```

Comment:
Internal functions should be in private headers.

Reply:
Where should the header be and what should it be called?

Reply:
Probably `include/types/wlr_pointer_constraints_v1.h`